third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,248 @@
+  //
+  // tensor shape: [128, 128]
+  // mfma layout: warpsPerCTA = [1, 4], instrShape = [32, 32]
+  // blocked layout: sizePerThread = [1, 4], threadsPerWarp = [32, 2],
+  // warpsPerCTA = [4, 1]
+  //
+  // minimal mfma tile is: [1*32, 4*32] = [32, 128]
+  // minimal blocked tile is: [1*32*4, 4*2*1] = [128, 8]
+  //
+  // Roughtly scratch buffer shape for conversion is:
+  // [max(32, 128), max(128, 16)] = [128, 128].
+  //
+  // This shape could be reduces by introducing intermediate
+  // layout and replacing old convert operations with two new conversions:
+  //
+  // %1 = convert %0 (mfma -> blocked)
+  //     ->
+  // %1 = convert %0 (mfma -> tmp)
+  // %2 = convert %1 (tmp -> blocked)
+  //
+  // Let's consider tmp as blocked layout:
+  // sizePerThread = [1, 4], threadsPerWarp = [32, 2], warpsPerCTA = [1, 4]
+  // Tmp layout scratch buffer has shape: [1*32*1, 4*2*4] = [32, 32]
+  //
+  // With intermediate layout we have two scratch buffers:
+  //
+  // %1 = convert %0 (mfma -> tmp): [max(32, 32), max(128, 32)] = [32, 128]
+  // %2 = convert %1 (tmp -> blocked): [max(32, 128), max(32, 32)] = [128, 32]
+  //
+  // Both of these buffers are 4x times smaller than original one and their live
+  // times do not intersect, therefore this transformation lowers LDS
+  // consumption.
+  void tryFitCvtIntoLDS(triton::gpu::ConvertLayoutOp cvtOp, int targetLDSSize) {
+    OpBuilder builder(cvtOp);
+
+    auto srcType = cvtOp.getSrc().getType();
+    auto dstType = cvtOp.getType();
+
+    auto srcEnc = srcType.getEncoding();
+    auto dstEnc = dstType.getEncoding();
+
+    auto ctx = srcEnc.getContext();
+    auto rank = srcType.getShape().size();
+    unsigned numWarps = triton::gpu::getNumWarpsPerCTA(srcEnc);
+    auto warpSize = triton::gpu::getWarpSize(srcEnc);
+
+    // Find all possible shapes of WarpsPerCTA by finding all possible
+    // factorizations of numWarps. Pick shape for which both conversions in
+    // decomposition use LDS less than LDSSize and for which sum of LDS usage
+    // is minimal. If no such shape exists, do not decompose.
+    auto factorizedNumWarps =
+        mlir::triton::AMD::factorizePowerOf2(numWarps, rank);
+    // Create a list of temporary layouts
+    SmallVector<Attribute> tmpLayouts;
+    for (int i = 0; i < factorizedNumWarps.size(); i++) {
+      auto warpsPerCTA = factorizedNumWarps[i];
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(srcEnc, warpsPerCTA));
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(dstEnc, warpsPerCTA));
+      SmallVector<unsigned> elemsPerThread(rank, 1);
+      SmallVector<unsigned> threadsPerWarp(rank, 1);
+      threadsPerWarp[rank - 1] = warpSize / 8;
+      threadsPerWarp[rank - 2] = warpSize / threadsPerWarp[rank - 1];
+      auto order = triton::gpu::getOrder(srcEnc);
+      auto layoutCTA = triton::gpu::getCTALayout(srcEnc);
+      auto fallbackLayout = triton::gpu::BlockedEncodingAttr::get(
```

Comment:
>2. Why is 8 chosen in warpSize / 8

For wave64 it will be [8, 8], for wave32 it will be [4, 8]. This is done to make layout tile "square", so no dim size of minimal tile is dominating.

> 3. In general, why we need this fallbackLayout? Is it covered by either srcEnc or dstEnc?

In some cases different warpsPerCTA of src or dst layout is not enough to reduce LDS usage, but some other layouts can be appropriate. These fallback layouts are designed to have as compact tile as possible, i.e. `elementsPerThread = [1, ... 1]`, and `threadsPerWarp` are as "square" as possible.

I believe, that in most cases fallback layout will be chosen as a temporary layout. This could be non optimal in terms of performance, but it is fine, because without this transformation kernel will not compile at all.